samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -19,36 +19,168 @@
+      List<StreamSpec> intStreams = plan.getIntermediateStreams();
+      if (!intStreams.isEmpty()) {
+        if (coordination != null) {
+          Latch initLatch = coordination.getLatch(1, LATCH_INIT);
+          coordination.getLeaderElector().tryBecomeLeader(() -> {
+            getStreamManager().createStreams(intStreams);
+            initLatch.countDown();
+          });
+          initLatch.await(LATCH_TIMEOUT, TimeUnit.MINUTES);
+        } else {
+          // each application process will try creating the streams, which
+          // requires stream creation to be idempotent
+          getStreamManager().createStreams(intStreams);
+        }
+      }
+
+      // 3. start the StreamProcessors
+      plan.getJobConfigs().forEach(jobConfig -> {
+        log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+
+        Object taskFactory = TaskFactoryUtil.createTaskFactory(config, app, this);
+        StreamProcessor processor = createStreamProcessor(jobConfig, taskFactory);
+        processor.addLifeCycleAware(new LocalProcessorListener(processor));
+        processors.add(processor);
+        processor.start();
+      });
+      appStatus = ApplicationStatus.Running;
+
+      // 4. block until the processors are done or there is a failure
+      latch.await();
+      if (throwable.get() != null) {
+        appStatus = ApplicationStatus.UnsuccessfulFinish;
+        throw throwable.get();
+      } else {
+        appStatus = ApplicationStatus.SuccessfulFinish;
+      }
+
+    } catch (Throwable t) {
+      throw new SamzaException("Failed to run application", t);
+    } finally {
+      coordination.reset();
+    }
   }
 
   @Override
   public void kill(StreamApplication streamApp) {
+    processors.forEach(StreamProcessor::stop);
   }
 
   @Override
   public ApplicationStatus status(StreamApplication streamApp) {
-    return null;
+    return appStatus;
   }
+
+  private CoordinationUtils getCoordinationUtils() throws Exception {
+    ApplicationConfig appConfig = new ApplicationConfig(config);
+    String clazz = appConfig.getCoordinationServiceFactoryClass();
+    if (clazz != null) {
+      CoordinationServiceFactory factory = ClassLoaderHelper.fromClassName(clazz);
+      String groupId = String.format("app-%s-%s", config.get(JobConfig.JOB_NAME()), config.get(JobConfig.JOB_ID(), "1"));
+      return factory.getCoordinationService(groupId, String.valueOf(processorId), config);
+    }
+    else {
+      return null;
+    }
+  }
+
+  private StreamProcessor createStreamProcessor(Config config, Object taskFactory) {
+    if (taskFactory instanceof StreamTaskFactory) {
+      return new StreamProcessor(processorId, config, new HashMap<>(), (StreamTaskFactory) taskFactory);
+    } else if (taskFactory instanceof AsyncStreamTaskFactory) {
+      return new StreamProcessor(processorId, config, new HashMap<>(), (AsyncStreamTaskFactory) taskFactory);
+    } else {
```

Comment:
What's the point of the factory pattern if StreamProcessor only supports 2 implementations? Maybe this is a question for @navina 

Ideally this createStreamProcessor method wouldn't even exist. We'd just pass any task factory to the StreamProcessor.